### PR TITLE
Add Open GL ES 3 support

### PR DESCRIPTION
**Pull request by orefkov**
Add support for GLES3 rendering API and it features:
- Instancing
- Uniform buffer object
- Multiple render targets
- Texture2DArray
- Texture3D
- Support for textures: ETC2, sRGB, floating point, depth, 1 & 2 channel (R & R/G).
Worked render pathes - Deferred, DeferredHWDepth, Prepass, PrepassHWDepth.

### CODE REVIEW

**Comment by weitjong** [comment]
Thanks for your PR. Could you squash the last few commits before we begin reviewing it.

**Comment by orefkov** [comment]
Done.

**Comment by orefkov** [comment]
Please run CI tests again - it fail on internal error on remote server.

**Comment by weitjong** [comment]
Actually you don’t need to worry about a lone CI build failure. We can tell if it is just technical.

File: Source/Urho3D/Graphics/Graphics.h [hunk 1]
```diff
@@ -795,6 +799,8 @@ class URHO3D_API Graphics : public Object
     String orientations_;
     /// Graphics API name.
     String apiName_;
+    String rendererName_;
+    String versionString_;
```
Comment (rokups) [inline Source/Urho3D/Graphics/Graphics.h]: These do not have doc comments. Yes, i realize they will be redundant. But for the sake of consistency..

File: Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp [hunk 2]
```diff
@@ -945,16 +967,20 @@ void Graphics::DrawInstanced(PrimitiveType type, unsigned indexStart, unsigned i
     glDrawElementsInstancedANGLE(glPrimitiveType, indexCount, indexType, reinterpret_cast<const GLvoid*>(indexStart * indexSize),
         instanceCount);
 #else
+  #ifndef GL_ES_VERSION_3_0
```
Comment (rokups) [inline Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp]: Preprocessor directives should be at the start of the line (in other places as well).

File: Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp [hunk 3]
```diff
@@ -2682,20 +2720,27 @@ unsigned Graphics::GetFloat32Format()
 
 unsigned Graphics::GetLinearDepthFormat()
 {
-#ifndef GL_ES_VERSION_2_0
+#ifndef GL_ES_VERSION_2_0
```
Comment (rokups) [inline Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp]: Trailing space. :eyes:

File: Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp [hunk 4]
```diff
@@ -2682,20 +2720,27 @@ unsigned Graphics::GetFloat32Format()
...
 {
-#ifndef GL_ES_VERSION_2_0
+#ifndef GL_ES_VERSION_2_0 
     // OpenGL 3 can use different color attachment formats
     if (gl3Support)
         return GL_R32F;
+    else
 #endif
+#ifdef GL_ES_VERSION_3_0
+    return GL_R16F;
```
Comment (rokups) [inline Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp]: This statement has wrong indentation in relation to previous `else`. Seems to me that removing `else` would be enough to maintain same behavior while not having confusing indentation.

File: Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp [hunk 5]
```diff
@@ -3303,7 +3377,7 @@ void Graphics::BindFramebuffer(unsigned fbo)
         glBindFramebufferEXT(GL_FRAMEBUFFER_EXT, fbo);
     else
 #endif
-        glBindFramebuffer(GL_FRAMEBUFFER, fbo);
+    glBindFramebuffer(GL_FRAMEBUFFER, fbo);
```
Comment (rokups) [inline Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp]: This indentation is also confusing.

File: Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp [hunk 6]
```diff
@@ -3382,21 +3456,25 @@ bool Graphics::CheckFramebuffer()
         return glCheckFramebufferStatusEXT(GL_FRAMEBUFFER_EXT) == GL_FRAMEBUFFER_COMPLETE_EXT;
     else
 #endif
-        return glCheckFramebufferStatus(GL_FRAMEBUFFER) == GL_FRAMEBUFFER_COMPLETE;
```
Comment (rokups) [inline Source/Urho3D/Graphics/OpenGL/OGLGraphics.cpp]: And this one.

File: Source/Urho3D/Graphics/OpenGL/OGLShaderVariation.cpp [hunk 7]
```diff
@@ -126,7 +126,18 @@ bool ShaderVariation::Create()
...
+#if defined(MOBILE_GRAPHICS) || URHO3D_GLES3
+        shaderCode += "#version 300 es\n";
+#else
         shaderCode += "#version 150\n";
+#endif
+    }
+#if defined(DESKTOP_GRAPHICS)
+    shaderCode += "#define DESKTOP_GRAPHICS\n";
+#elif defined(MOBILE_GRAPHICS)
+    shaderCode += "#define MOBILE_GRAPHICS\n";
```
Comment (rokups) [inline Source/Urho3D/Graphics/OpenGL/OGLShaderVariation.cpp]: These are very useful defines :+1:

File: bin/CoreData/Shaders/GLSL/PostProcess.glsl [hunk 8]
```diff
@@ -64,7 +64,7 @@ vec3 Uncharted2Tonemap(vec3 x)
    return ((x*(A*x+C*B)+D*E)/(x*(A*x+B)+D*F))-E/F;
 }
 
-#ifndef GL_ES
+#if !defined(GL_ES) || __VERSION >= 300
```
Comment (rokups) [inline bin/CoreData/Shaders/GLSL/PostProcess.glsl]: Typo in `__VERSION`?

File: bin/CoreData/Shaders/GLSL/Samplers.glsl [hunk 9]
```diff
@@ -9,21 +9,21 @@ uniform samplerCube sEnvCubeMap;
 uniform sampler2D sLightRampMap;
 uniform sampler2D sLightSpotMap;
 uniform samplerCube sLightCubeMap;
-#ifndef GL_ES
-    uniform sampler3D sVolumeMap;
+#if !defined(GL_ES) || __VERSION__>=300
```
Comment (rokups) [inline bin/CoreData/Shaders/GLSL/Samplers.glsl]: Spaces around operators missing.

**Comment by rokups** [comment]
This PR is simply awesome. I left bunch of comments pointing out cosmetic stuff, but that is minor. @orefkov thanks for such a great PR :)

**Comment by orefkov** [comment]
I wil fix this cosmetic stuff and commit it ASAP.

**Comment by weitjong** [comment]
Sorry for the delay. Your PR is being CI build tested across all the platforms now and we will see from there.

**Comment by weitjong** [comment]
What is your rationale to replace the standard define "GL_ES_VERSION_2_0" with Urho3D-specific "URHO3D_GLES2"? I would prefer to keep the standard define if there is no good reason in doing so.

**Comment by orefkov** [comment]
Due to backward compatibility of GLES3 with GLES2, the macro GL_ES_VERSION_2_0 is also defined when connecting header files from GLES3. That is, by this definition it is impossible to determine whether we use GLES2 or GLES3. Therefore, I added the symbol URHO3D_GLES2, which is defined when assembling Urho3D for GLES2.

**Comment by weitjong** [comment]
I see. Thanks for your explanation. I am trying to make a few minor changes and therefore need to understand the rationale in your proposed changes. I think Lasse’s original code is easy to read and reason about because he commented the code well.

**Comment by weitjong** [comment]
I don't see any changes on the SDL side. AFAIK SDL only supports GLES 2. Do you recreate the GLES context or the GLES 2 context from SDL works just fine?

I also notice that the new build option from CI environment variable was not passed correctly to the CI build job. After fixing this, I expect at least ARM CI build to fail because the ARM sysroot that we have currently does not have ```libGLESv3.so``` yet.

**Comment by orefkov** [comment]
In Graphics::SetMode
        SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
worked fine.
libGLESv3.so linked only for Android, for ARM URHO3D_GLES3 was set to FALSE little above.

**Comment by weitjong** [comment]
I don't have the generic ARM board to verify my build, but it appears I am able to do 32-bit ARM build with URHO3D_GLES 3 enabled using `libgles2-mesa-dev` after a minor tweak on the CMake build script (basically compiled using GLES3 headers, but linked using libGLESv2.so). If you are OK, I will push my commits on top of yours in the dev branch.

**Comment by orefkov** [comment]
Im OK :)

**Comment by weitjong** [comment]
Have anyone tried to do Web build on the dev branch (https://github.com/urho3d/Urho3D/tree/orefkov-GLES3_Support)? I have encountered a runtime error. On the Crome console I got:

```
06_SkeletalAnimation.html:1 [Sat Nov 23 22:56:58 2019] ERROR: Failed to compile pixel shader Shadow():
06_SkeletalAnimation.html:1 ERROR: 0:59: '' : No precision specified for (float)
printErr @ 06_SkeletalAnimation.html:1
06_SkeletalAnimation.html:1 ERROR: 0:60: '' : No precision specified for (float)
printErr @ 06_SkeletalAnimation.html:1
06_SkeletalAnimation.html:1 ERROR: 0:61: '' : No precision specified for (float)
```

This was still with URHO3D_GLES3 set to disabled, so it is a regression issue.

I am working on a commit to bring the GLES 3 / WebGL 2 available for the Web build and actually managed to have a clean build. However, I got the above similar runtime error too. Fortunately (or unfortunately) I realized it was caused by the regression issue.

**Comment by weitjong** [comment]
Can someone confirm my observation that this PR has regression issue with Web build? Thanks.

**Comment by eugeneko** [comment]
Does our CI store build artifacts that can be downloaded?

чт, 16 янв. 2020 г. в 18:00, Yao Wei Tjong <notifications@github.com>:

**Comment by weitjong** [comment]
Currently the CI only does that when the build is triggered by a commit in the master branch.

**Comment by SirNate0** [comment]
I had to update cmake and emscripten (I believe it was > 3.14 and > 1.39.1 respectively if anyone else is going to try). . I didn't pass any specific flags, but with a clean web build I am also getting a large number of `ERROR: 0:61: '' : No precision specified for (float)` type errors with emscripten 1.39.6. If I needed to specify something to use the right GLES version let me know and I'll do another build.

Also, for a while now cmake_emscripten.sh has not been finding the emscripten compilers after running `emsdk activate {version}` and `. emsdk_env.sh` unless I explicitly set the environment variable `EMSCRIPTEN_ROOT_PATH=/home/git/tools/emsdk/upstream/emscripten` when calling the script even though emcc and the like are in the PATH.

**Comment by weitjong** [comment]
Yes, that’s the error I got too. Thanks for confirmation. One of proposed change is the culprit. See my post in the forum.

Regarding EMSCRIPTEN_ROOT_PATH, I see what I can do later.

**Comment by weitjong** [comment]
Just have a quick look. It could be done easily, at least on *nix host. It should be quite safe to rely on EMSDK internal configuration file to determine the location of this path.

**Comment by pengfei666** [comment]
I am a little bit confused about the PR. How about if a device doesn't support GLES 3？ 

I don't think using a macro `URHO3D_GLES3` is good. Actually, I think it should be a runtime variable instead of a compile time constant variable. If a device doesn't support GLES3, it can fallback to GLES2 automatically. 

so the code should be
```

if(glsl3_is_supported)
{
  // do something
}
else
{
  // do something
}
```

**Comment by kakashidinho** [comment]
Hi, I’m interested in when this PR will be merged?

**Comment by weitjong** [comment]
From what I recall, I have commented last that there is a regression issue with the Web platform. So, unless OP or someone else willing to help to fix that, this PR won't be merged just yet.

File: Source/Urho3D/Graphics/OpenGL/OGLShaderVariation.cpp [hunk 10]
```diff
@@ -126,7 +126,18 @@ bool ShaderVariation::Create()
     }
     // Force GLSL version 150 if no version define and GL3 is being used
     if (!verEnd && Graphics::GetGL3Support())
+    {
+#ifdef GL_ES_VERSION_3_0
+        shaderCode += "#version 300 es\n";
```
Comment (AntonioDLC) [inline Source/Urho3D/Graphics/OpenGL/OGLShaderVariation.cpp:131]:
```suggestion
        shaderCode += "#version 300 es\n" 
        
        //Float precision is mandatory for OpenGL ES Shader Language 3.00
        shaderCode +="precision highp float;\n";

```

**Review by AntonioDLC — CHANGES_REQUESTED** [review]
I have never meddled with this engine nor sanity checked this fix, but since I'm willing to bet this little addition should get the PR going finally free of the float precision error mentioned above.

**Comment by 1vanK** [comment]
Хотя на линуксе вроде можно запускать приложения в режиме GL ES, надо будет покопаться

**Comment by orefkov** [comment]
> У нас щас примеры на андроиде пустое окно окно показывают (лаунчер их походу не находит их, мб дело в регекспе). Как тестить я хз. Мобилки вне сферы моих интересов, если ты скажешь что можно смержить, я смержу

Хм похоже я какие-то большие нововведения пропустил :)
Впервые за три года решил тут покопаться немного, и на те, такие новости.
Что за сыр-бор с русскоязычностью?

По этому PR'у - у меня у самого только одна мобильная игрушка, она работает. Но там половина движка не включена при конфиге билда. Я постараюсь в ближайшее время собрать полный движок с примерами на андроид и проверить.
Отпишусь, как получится.
И на свежем emscipten не может никак собраться, и под виндой пробовал, и под линуксом...

**Comment by 1vanK** [comment]
`#ifdef URHO3D_OPENGL`

Обрати внимание, что сейчас могут быть оба API доступны, т.е. одновременно определены `URHO3D_D3D11` и `URHO3D_OPENGL`. В DebugHud такой дефайн вижу

Т.е. помимо дефайна все что за пределами папок OpenGL и Direct3D11 имеет более сложную проверку

Пример:

```
#ifdef URHO3D_OPENGL
    if (gapi == GAPI_OPENGL)
    {
        Constructor_OGL();
        return;
    }
#endif

#ifdef URHO3D_D3D11
    if (gapi == GAPI_D3D11)
    {
        Constructor_D3D11();
        return;
    }
#endif
```

**Comment by 1vanK** [comment]
Игра не компилируется или движок не компилируется? Я тестил движок недавно все через пень колоду но собирается, ты можешь процесс сборки под линукс глянуть тут: https://github.com/urho3d/Urho3D/blob/master/.github/workflows/main.yml#L333-L347

а под виндой у меня батник для создания проектов

```
:: При запуске батника от Админа меняется текущий путь. Фиксим это.
cd /D "%~dp0"

::set "PATH=c:\Programs\CMake\bin;c:\Programs\Doxygen;c:\Programs\Graphviz\bin;c:\Program Files (x86)\HTML Help Workshop;c:\Program Files\Git\bin;c:\Programs\Python3;c:\Programs\MinGW64\i686-8.1.0-posix-dwarf-rt_v6-rev0\mingw32\bin"

set "PATH=c:\Programs\CMake\bin;c:\Programs\Doxygen;c:\Programs\Graphviz\bin;c:\Program Files (x86)\HTML Help Workshop;c:\Program Files\Git\bin;c:\Programs\Python3;PATH=c:\Programs\CMake\bin;c:\Programs\Doxygen\;c:\Programs\Graphviz\bin\;c:\Program Files (x86)\HTML Help Workshop\;c:\msys64\mingw64\bin\"



call g:\MyGames\Emscripten\emsdk\emsdk_env.bat

::set "EMSCRIPTEN=g:\MyGames\Emscripten\emsdk"

::set "TOOLCHAINS=%~dp0repo\cmake\Toolchains"
set "TOOLCHAINS=g:\MyGames\Emscripten\emsdk\upstream\emscripten\cmake\Modules\Platform"

::mkdir BuildEmscripten
::cd BuildEmscripten

::cmake.exe ../Urho3D -DWEB=1 -DCMAKE_TOOLCHAIN_FILE="%TOOLCHAINS%\Emscripten.cmake" -DURHO3D_OPENGL=1 -DURHO3D_ANGELSCRIPT=1 -DURHO3D_LUA=1 -DURHO3D_NETWORK=1 -DURHO3D_STATIC_RUNTIME=1 -DURHO3D_SAMPLES=1 -DURHO3D_TOOLS=0 -DURHO3D_DOCS=0 -DURHO3D_LIB_TYPE=SHARED -DURHO3D_DATABASE_ODBC=0 -DURHO3D_GENERATEBINDINGS=0 -DURHO3D_TRACY_PROFILING=1

::cd repo
::call script\cmake_emscripten.bat g:\MyGames\urho3d_fork\BuildEmscripten -D URHO3D_PCH=0


::cmake.exe repo -B build_web -G "MinGW Makefiles" -D WEB=1 -D CMAKE_TOOLCHAIN_FILE="%TOOLCHAINS%\Emscripten.cmake" -DURHO3D_OPENGL=1 -DURHO3D_ANGELSCRIPT=1 -DURHO3D_LUA=1 -DURHO3D_NETWORK=1 -DURHO3D_STATIC_RUNTIME=1 -DURHO3D_SAMPLES=1 -DURHO3D_TOOLS=0 -DURHO3D_DOCS=0 -DURHO3D_LIB_TYPE=SHARED -DURHO3D_DATABASE_ODBC=0 -DURHO3D_GENERATEBINDINGS=0 -DURHO3D_TRACY_PROFILING=0

cmake.exe repo -B build_web -G "MinGW Makefiles" -D WEB=1 -D CMAKE_TOOLCHAIN_FILE="%TOOLCHAINS%\Emscripten.cmake" -D EMSCRIPTEN_SHARE_DATA=1

::-D MEMORY64=1 

pause
```

и для компиляции 

```
:: При запуске батника от Админа меняется текущий путь. Фиксим это.
cd /D "%~dp0"

::set "PATH=c:\Programs\CMake\bin;c:\Programs\Doxygen;c:\Programs\Graphviz\bin;c:\Program Files (x86)\HTML Help Workshop;c:\Program Files\Git\bin;c:\Programs\Python3;c:\Programs\MinGW64\i686-8.1.0-posix-dwarf-rt_v6-rev0\mingw32\bin"
set "PATH=c:\Programs\CMake\bin;c:\Programs\Doxygen;c:\Programs\Graphviz\bin;c:\Program Files (x86)\HTML Help Workshop;c:\Program Files\Git\bin;c:\Programs\Python3;PATH=c:\Programs\CMake\bin;c:\Programs\Doxygen\;c:\Programs\Graphviz\bin\;c:\Program Files (x86)\HTML Help Workshop\;c:\msys64\mingw32\bin\"

call g:\MyGames\Emscripten\emsdk\emsdk_env.bat

::set "TOOLCHAINS=%~dp0Urho3D\cmake\Toolchains"

::cmake --build build_web --target Urho3DPlayer >log 2>&1
cmake --build build_web

pause
```

**Comment by orefkov** [comment]
На emscripten 3.1.17 с правильно указанным toolchain собралось.
Пофиксил сборки для WebGL - проверил, работают все варианты - Android GLES2, Android GLES3, Web GLES2, Web GLES3.
Для сборки с GLES3 надо задавать опцию URHO3D_GLES3, по дефолту она FALSE.

**Comment by 1vanK** [comment]
Ну я волью тогда как есть, потом фиксанем если что?

**Comment by 1vanK** [comment]
А вообще статистика устройств opengl es 2 есть какая-то? Имеет смысл тащить его дальше? Я вот тут смотрю, Web GL тоже на ES 3 перешел https://en.wikipedia.org/wiki/WebGL

**Comment by 1vanK** [comment]
Ну хотя оно пока никак не мешает

**Comment by orefkov** [comment]
> А вообще статистика устройств opengl es 2 есть какая-то? Имеет смысл тащить его дальше? Я вот тут смотрю, Web GL тоже на ES 3 перешел https://en.wikipedia.org/wiki/WebGL

Согласно https://developer.android.com/about/dashboards#OpenGL на GLES2 осталось 6.77% устройств.
И я думаю, вряд ли это устройства геймеров :) Так что есть все резоны сделать GLES3 дефолтом для Android и WebGL.

**Comment by 1vanK** [comment]
Кстати, когда на мобилке приложение запускается, куда лог пишется? Дёргается функция `SDL_GetPrefPath()`. В винде она ведет в `c:\Users\User\AppData\Roaming\urho3d\`, а на Андроиде куда?

**Comment by orefkov** [comment]
Вроде /data/data/имяпакета/files

чт, 22 дек. 2022 г., 13:53 1vanK ***@***.***>:

**Comment by 1vanK** [comment]
Странная фигня. Спрайт батч в браузере черный прямоугольник рендерит вместо спрайта. Вычитал, что для GL ES 1.0 ограничение, что текстуры должны иметь размер степени двойки (64, 128 и так далее), проверил и правда,  "правильные" текстуры рендерит. Но при этом есть пример `51_Urho2DStretchableSprite`, где используется текстура 200x200, и она без проблем рендерится в вебе. Но не нашел, где размер текстуры корректируется до степени двойки. Просто текстура через ResourceCache загружается, как и я в спрайт баче делаю

**Comment by 1vanK** [comment]
Забавно, если текстуру загрузить вот таким вот экзотическим способом (через спрайт2d), то она нормально работает даже в вебе:

```
auto* sprite = cache->GetResource<Sprite2D>("Urho2D/Stretchable.png");
Texture2D* texture= sprite->GetTexture();
```

**Pull request merged**